llm/server.go

```diff
@@ -217,15 +218,93 @@ func NewLlamaServer(gpus discover.GpuInfoList, model string, ggml *GGML, adapter
 		params = append(params, "--threads", strconv.Itoa(defaultThreads))
 	}
 
-	flashAttnEnabled := envconfig.FlashAttention()
+	isEmbeddingModel := false
+	if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+		isEmbeddingModel = true
+	}
+
+	setCacheTypeParam := func(paramName, cacheType string) {
+		if cacheType == "" {
+			return
+		}
+
+		validCacheTypes := []string{"f32", "f16", "q8_0", "q5_1", "q5_0", "iq4_nl", "q4_1", "q4_0"}
+		if !slices.Contains(validCacheTypes, cacheType) {
+			slog.Warn("invalid cache type, ignoring", "param", paramName, "type", cacheType)
+			return
+		}
+
+		// For embedding models, only allow f16 and f32
+		if isEmbeddingModel && cacheType != "f16" && cacheType != "f32" {
+			slog.Warn("only f16 and f32 cache types are supported for embedding models, ignoring",
+				"param", paramName, "type", cacheType)
+			return
+		}
+
+		params = append(params, paramName, cacheType)
+		slog.Debug("Setting cache type", "param", paramName, "type", cacheType)
+	}
+
+	// Define cacheTypeK and cacheTypeV
+	cacheTypeK := envconfig.CacheTypeK()
+	cacheTypeV := envconfig.CacheTypeV()
+
+	// Set cache types only if they are not empty
+	supportsFlashAttention := func(ggml *GGML) bool {
```

Comment:
Is this check comprehensive (i.e. is it architecture based too?)

Reply:
Improved code comments around this function.

Reply:
Got it, wasn't sure if there was a set of models we should include (or omit) from the check. Thanks